sway-core/src/asm_generation/fuel/optimizations.rs

```diff
@@ -57,145 +57,184 @@ impl AbstractInstructionSet {
+                                if get_def_version(&latest_version, &base_reg.reg)
+                                    == base_reg.ver
+                                    && ((offset / 8) + imm.value as u64)
+                                        < compiler_constants::TWELVE_BITS =>
+                            {
+                                // bail if LW cannot read where this memory is
+                                if offset % 8 == 0 {
+                                    let new_imm = VirtualImmediate12::new_unchecked(
+                                        (offset / 8) + imm.value as u64,
+                                        "Immediate offset too big for LW",
+                                    );
+                                    let new_lw =
+                                        VirtualOp::LW(dest.clone(), base_reg.reg.clone(), new_imm);
+                                    // The register defined is no more useful for us. Forget anything from its past.
+                                    reg_contents.remove(dest);
+                                    record_new_def(&mut latest_version, dest);
+                                    // Replace the LW with a new one in-place.
+                                    *op = new_lw;
+                                }
+                            }
+                            _ => {
+                                reg_contents.remove(dest);
+                                record_new_def(&mut latest_version, dest);
+                            }
+                        },
+                        VirtualOp::SW(addr_reg, src, imm) => match reg_contents.get(addr_reg) {
+                            Some(RegContents::BaseOffset(base_reg, offset))
+                                if get_def_version(&latest_version, &base_reg.reg)
+                                    == base_reg.ver
+                                    && ((offset / 8) + imm.value as u64)
+                                        < compiler_constants::TWELVE_BITS =>
+                            {
                                 let new_imm = VirtualImmediate12::new_unchecked(
                                     (offset / 8) + imm.value as u64,
-                                    "Immediate offset too big for LW",
+                                    "Immediate offset too big for SW",
+                                );
+                                let new_sw =
+                                    VirtualOp::SW(base_reg.reg.clone(), src.clone(), new_imm);
+                                // Replace the SW with a new one in-place.
+                                *op = new_sw;
+                            }
+                            _ => (),
+                        },
+                        VirtualOp::MOVE(dest, src) => {
+                            let ver = get_def_version(&latest_version, src);
+                            if let Some(RegContents::BaseOffset(src, 0)) = reg_contents.get(src) {
```

Comment:
are you trying to catch something like
```
a = b
b = a
```
?? Not sure I understand what is it that you're trying to optimize, and why the existing copy propagation in the assembly generator doesn't apply.

Reply:
For example, from the `configurable_const` test:
![image](https://github.com/user-attachments/assets/3d9c7bae-270f-4009-b323-d6c21d1ab7ee)

Reply:
On `a = b`, this match arm does two things:
1 - first, it indeed checks if `a` and `b` are already the same, and it removes the instruction;
2 - but it also propagates that `b` and `a` are the same for future optimizations.